rollouts/pkg/packagediscovery/packagediscovery.go

```diff
@@ -58,83 +134,118 @@ func NewPackageDiscovery(client client.Client, namespace string) *PackageDiscove
+	for _, repo := range repos {
+		repoPackages, err := d.getPackagesForRepo(ghc, ctx, gitHubSelector, repo)
 		if err != nil {
 			return nil, fmt.Errorf("unable to get packages: %w", err)
 		}
-
 		discoveredPackages = append(discoveredPackages, repoPackages...)
 	}
-
-	d.cache = &Cache{
-		packages:   discoveredPackages,
-		config:     config,
-		expiration: time.Now().Add(1 * time.Minute),
-	}
-
 	return discoveredPackages, nil
 }
 
-func (d *PackageDiscovery) getRepositoryNames(gitHubClient *github.Client, selector gitopsv1alpha1.GitHubSelector, ctx context.Context) ([]string, error) {
-	repositoryNames := []string{}
+func (d *PackageDiscovery) getRepos(gitHubClient *github.Client, selector gitopsv1alpha1.GitHubSelector, ctx context.Context) ([]*github.Repository, error) {
+	var matchingRepos []*github.Repository
 
 	if isSelectorField(selector.Repo) {
 		// TOOD: add pagination
 		listOptions := github.RepositoryListOptions{}
 		listOptions.PerPage = 150
 
-		repositories, _, err := gitHubClient.Repositories.List(ctx, selector.Org, &listOptions)
+		repos, _, err := gitHubClient.Repositories.List(ctx, selector.Org, &listOptions)
 		if err != nil {
 			return nil, err
 		}
 
-		allRepositoryNames := []string{}
-		for _, repository := range repositories {
-			allRepositoryNames = append(allRepositoryNames, *repository.Name)
+		allRepoNames := []string{}
+		for _, repo := range repos {
+			allRepoNames = append(allRepoNames, *repo.Name)
 		}
 
-		matchRepositoryNames := filterByPattern(selector.Repo, allRepositoryNames)
+		matchingRepoNames := filterByPattern(selector.Repo, allRepoNames)
+		matchingRepoNameSet := sets.String{}
+		matchingRepoNameSet.Insert(matchingRepoNames...)
 
-		repositoryNames = append(repositoryNames, matchRepositoryNames...)
+		for _, repo := range repos {
+			if matchingRepoNameSet.Has(*repo.Name) {
+				matchingRepos = append(matchingRepos, repo)
+			}
+		}
 	} else {
-		repositoryNames = append(repositoryNames, selector.Repo)
+		repo, _, err := gitHubClient.Repositories.Get(ctx, selector.Org, selector.Repo)
+		if err != nil {
+			return nil, err
+		}
+		matchingRepos = append(matchingRepos, repo)
 	}
 
-	return repositoryNames, nil
+	return matchingRepos, nil
 }
 
-func (d *PackageDiscovery) getPackagesForRepository(gitHubClient *github.Client, ctx context.Context, selector gitopsv1alpha1.GitHubSelector, repoName string) ([]DiscoveredPackage, error) {
+func (d *PackageDiscovery) getPackagesForRepo(gitHubClient *github.Client, ctx context.Context, selector gitopsv1alpha1.GitHubSelector, repo *github.Repository) ([]DiscoveredPackage, error) {
 	discoveredPackages := []DiscoveredPackage{}
-
+	branch := selector.Branch
+	if branch == "" {
+		branch = repo.GetDefaultBranch()
+	}
 	if isSelectorField(selector.Directory) {
-		tree, _, err := gitHubClient.Git.GetTree(ctx, selector.Org, repoName, selector.Revision, true)
+		tree, _, err := gitHubClient.Git.GetTree(ctx, selector.Org, *repo.Name, repo.GetDefaultBranch(), true)
```

Comment:
Why are we using `repo.GetDefaultBranch` here instead of `branch`?

Reply:
Good catch, fixed it.